File: backend/app/models/student.py
"""
Modelo ORM para la tabla students.

Nota sobre JSONB: SQLModel no incluye soporte nativo para el tipo JSONB
de PostgreSQL. Se usa sa_column con Column(JSONB) de SQLAlchemy para el
campo periodos_academicos de modo que psycopg2 lo serialice correctamente.
"""

import uuid
from datetime import datetime
from typing import Any, Optional

from sqlalchemy import Column
from sqlalchemy.dialects.postgresql import JSONB
from sqlmodel import Field, SQLModel


class Student(SQLModel, table=True):
    __tablename__ = "students"

    # Identidad y autenticacion
    id: uuid.UUID = Field(
        default_factory=uuid.uuid4,
        primary_key=True,
    )
    email: str = Field(unique=True, index=True)
    password_hash: str

    # Datos del estudiante (de historial.estudiante)
    codigo_matricula: str = Field(unique=True, index=True)
    nombres_apellidos: str
    facultad: Optional[str] = None
    escuela: Optional[str] = None
    plan: Optional[str] = None

    # URLs de PDFs
    pdf_url_historial: Optional[str] = None
    pdf_url_matricula: Optional[str] = None
    pdf_url_cv: Optional[str] = None

    # Historial academico serializado (JSONB)
    periodos_academicos: Optional[Any] = Field(
        default=None,
        sa_column=Column(JSONB, nullable=False, server_default="'[]'"),
    )

    # CV en texto plano extraido por LLM
    cv_text: Optional[str] = None

    # Idea de Tesis inicial
    thesis_idea: Optional[str] = Field(default="")

    # Resumen de creditos
    creditaje_requerido_para_egresar: Optional[int] = None
    creditaje_aprobado: Optional[int] = None
    obligatorios: Optional[int] = None
    de_especialidad: Optional[int] = None
    electivos_generales: Optional[int] = None
    electivos_de_especialidad: Optional[int] = None
    optativos: Optional[int] = None
    alternativos: Optional[int] = None
    de_otra_especialidad: Optional[int] = None
    mas_de_una_vez: Optional[int] = None
    otros: Optional[int] = None
    creditaje_faltante: Optional[int] = None
    promedio_ponderado: Optional[float] = None

    # Timestamps
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)

File: frontend/src/types/student.ts
// Profile returned by the API after processing PDFs

export interface Course {
  ciclo: string;
  plan: string;
  tipo: string;
  codigo: string;
  asignatura: string;
  calificacion: string;
  creditos: string;
  seccion: string;
}

export interface AcademicPeriod {
  periodo: string;
  cursos: Course[];
}

export interface CreditsSummary {
  creditaje_requerido_para_egresar: number;
  creditaje_aprobado: number;
  obligatorios: number;
  de_especialidad: number;
  electivos_generales: number;
  electivos_de_especialidad: number;
  optativos: number;
  alternativos: number;
  de_otra_especialidad: number;
  mas_de_una_vez: number;
  otros: number;
  creditaje_faltante: number;
  promedio_ponderado: number;
}

export interface StudentInfo {
  codigo_matricula: string;
  nombres_apellidos: string;
  facultad: string;
  escuela: string;
  plan: string;
}

export interface StudentProfile {
  student_id: string;
  historial: {
    estudiante: StudentInfo;
    periodos_academicos: AcademicPeriod[];
    resumen_creditos: CreditsSummary;
  };
  cv?: {
    cv_text: string;
  };
  thesis_idea: string;
  pdf_urls: {
    historial?: string;
    matricula?: string;
    cv?: string;
  };
  elapsed_seconds?: number;
}

// Stored user session
export interface AuthUser {
  student_id: string;
  email: string;
  estudiante: StudentInfo;
  periodos_academicos: AcademicPeriod[];
  resumen_creditos: CreditsSummary;
  cv_text: string;
  thesis_idea: string;
  pdf_urls: {
    historial?: string;
    matricula?: string;
    cv?: string;
  };
}

// DB stats for the dashboard
export interface DbStats {
  advisors: number;
  theses: number;
  publications: number;
  range_start: number;
  range_end: number;
}
